### PR TITLE
fix(router) rewrite cause api match fail

### LINKED ISSUE #3140: rewrite cause api match fail
### Summary

I write a custom plugin to rewrite uri, the plugin is global, the plugin rewrite function is be called on every request, that's OK. But during the match api phase, something wrong. In router.lua, in function exec, it use ngx.var.request_uri to match api
```
    local method      = ngx.req.get_method()
    local request_uri = ngx.var.request_uri
    local uri         = request_uri

    do
      local idx = find(uri, "?", 2, true)
      if idx then
        uri = sub(uri, 1, idx - 1)
      end
    end

    local req_host = ngx.var.http_host

    local match_t = find_api(method, uri, req_host, ngx)
    if not match_t then
      return nil
    end
```
request_uri is a readonly variable ,can't be changed in rewrite phase.
### Steps To Reproduce

1. write a custom plugin to rewrite a uri
2. enable the plugin globally
3. make a request
4. rewrite is Ok, but can't match api

### Additional Details & Logs

- Kong version (0.11.0)
- CentOS6

### PR DESCRIPTION
**Pull request by samwonwyd**
In rewrite phase, we use ngx.req.set_uri() to change uri, but in
router.lua it use ngx.var.request_uri to match api, but this
variable is readonly, we can't change it in rewrite phase, and
also cause api match failed.

### Issues resolved

Fix #3140

### CODE REVIEW

**Comment by thibaultcha** [comment]
Thanks for the patch. Unfortunately this won’t do it I am afraid. We purposefully match against `request_uri` because `uri` normalizes the path and caused some unexpected behavior for some users, and our goal is to provide a transparent reverse-proxy. I am actually surprised (and worried) that this change is not causing a CI failure. 

Additionally, the removal of the querystring stripping change is not desired, we need this stripping for various reasons.

I have given this some thought in the past (but the use-case has rarely been requested) and I believe a better approach would be to allow for an `ngx.ctx` variable to be set, and if so, take precedence in the router over `request_uri`.

**Comment by thibaultcha** [comment]
Also, we need tests for any change that is being submitted. Will you provide tests for your patch? Thank you.

File: kong/core/router.lua [hunk 1]
```diff
@@ -821,15 +821,9 @@ function _M.new(apis)
 
   function self.exec(ngx)
     local method      = ngx.req.get_method()
-    local request_uri = ngx.var.request_uri
-    local uri         = request_uri
+    local uri         = ngx.req.uri
```
Comment (bungle) [inline kong/core/router.lua]: Does this really work? I think we previously had `ngx.var.uri` here. If there is `ngx.req.uri` and that is not badly `normalized` then that could open some possibilities.

Comment (bungle) [reply kong/core/router.lua]: I tested this with `openresty/1.11.2.4` and there is no such thing as `ngx.req.uri`.

**Comment by thibaultcha** [comment]
@samwonwyd Ping. Are you able to provide the changes we suggested? We might have to close this PR if it becomes inactive.

File: kong/core/router.lua (hunk 1 above)
Comment (samwonwyd) [reply kong/core/router.lua]: Sorry, my mistake.It's `ngx.var.uri`, there is no  `ngx.req.uri`.

Comment (thibaultcha) [reply kong/core/router.lua]: @samwonwyd Hi, nice to hear back from you! This shouldn't be a concern anyway (we do not want to revert this value to `ngx.var.uri` as both @bungle and I pointed out. Instead, will you be able to provide an implementation based on my above feedback?

Comment (samwonwyd) [reply kong/core/router.lua]: Ok, I'm trying to write the tests.

**Comment by thibaultcha** [comment]
Closing due to lack of updates. If you @samwonwyd or someone else is still interested in this feature, we are open to receiving a PR for it (granted the approach is cleaner, as debated in this thread). Thanks!

**Comment by Tieske** [comment]
here's an ugly hack I used to work around this;
```lua
--------------------------------------------------------------------------------
-- Ugly hack to work around issue https://github.com/Kong/kong/pull/3141
local put_hack_in_place, set_request_uri do
  local HACK_IN_PLACE = false
  local original__index

  function put_hack_in_place()
    if not HACK_IN_PLACE then
      HACK_IN_PLACE = true
      local mt = getmetatable(ngx.var)
      original__index = mt.__index
      mt.__index = function(self, key)
        if key == "request_uri" then
          return ngx.ctx.hacked_request_uri or original__index(self, key)
        end
        return original__index(self, key)
      end
    end
  end

  function set_request_uri(uri)
    ngx.ctx.hacked_request_uri = uri
  end
end
--------------------------------------------------------------------------------
```

**Pull request closed without merging**